File: 3-APIs/kaffebar-java-solution/src/main/java/no/soprasteria/kaffebar/controller/OrdersController.java
package no.soprasteria.kaffebar.controller;

import no.soprasteria.kaffebar.api.OrdersApi;
import no.soprasteria.kaffebar.error.NotFoundException;
import no.soprasteria.kaffebar.events.OrderEventPublisher;
import no.soprasteria.kaffebar.model.Coffee;
import no.soprasteria.kaffebar.model.CreateOrderRequest;
import no.soprasteria.kaffebar.model.Order;
import no.soprasteria.kaffebar.model.OrderStatus;
import no.soprasteria.kaffebar.model.UpdateOrderStatusRequest;
import no.soprasteria.kaffebar.store.Menu;
import no.soprasteria.kaffebar.store.OrderStore;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.RestController;

import java.net.URI;
import java.util.List;
import java.util.UUID;

/**
 * Implementerer det genererte {@code OrdersApi}-interfacet — oppgave 1, 5, 7 og 8.
 *
 * Legg merke til hvor lite som står her. Ruting, media types, binding av
 * query-parametere med default-verdier, og all validering kommer fra kontrakten via
 * det genererte interfacet. Det som er igjen er forretningslogikken. Det er poenget
 * med contract-first.
 */
@RestController
public class OrdersController implements OrdersApi {

    private final OrderStore store;
    private final OrderEventPublisher events;

    public OrdersController(OrderStore store, OrderEventPublisher events) {
        this.store = store;
        this.events = events;
    }

    /** Oppgave 8. limit og offset har allerede fått default-verdiene sine fra kontrakten. */
    @Override
    public ResponseEntity<List<Order>> listOrders(OrderStatus status, Integer limit, Integer offset) {
        return ResponseEntity.ok(store.list(status, limit, offset));
    }

    /** Oppgave 1. 201 med Location-header — det er det `Created` betyr. */
    @Override
    public ResponseEntity<Order> createOrder(CreateOrderRequest request) {
        Coffee coffee = Menu.find(request.getCoffeeId())
                // 404 og ikke 400: kontrakten er overholdt, ressursen finnes bare ikke.
                .orElseThrow(() -> new NotFoundException(
                        "Fant ingen kaffe med id " + request.getCoffeeId() + ". Se GET /menu."));

        Order order = store.create(request, coffee);
        events.orderCreated(order);

        return ResponseEntity
                .created(URI.create("/orders/" + order.getOrderId()))
                .body(order);
    }

    /** Oppgave 5 + 6. */
    @Override
    public ResponseEntity<Order> getOrder(UUID orderId) {
        Order order = store.find(orderId).orElseThrow(() -> notFound(orderId));
        return ResponseEntity.ok(order);
    }

    /** Oppgave 7, anbefalt variant. */
    @Override
    public ResponseEntity<Order> updateOrderStatus(UUID orderId, UpdateOrderStatusRequest request) {
        return ResponseEntity.ok(applyStatus(orderId, request.getStatus()));
    }

    /**
     * Oppgave 7, action-varianten. Samme implementasjon, to endepunkter — se FASIT.md
     * for hvorfor begge finnes og hvorfor PATCH er anbefalingen.
     */
    @Override
    public ResponseEntity<Order> setOrderStatus(UUID orderId, UpdateOrderStatusRequest request) {
        return ResponseEntity.ok(applyStatus(orderId, request.getStatus()));
    }

    private Order applyStatus(UUID orderId, OrderStatus status) {
        OrderStore.StatusChange change = store.setStatus(orderId, status)
                .orElseThrow(() -> notFound(orderId));

        // Publiser bare når noe faktisk endret seg. Ellers ville en frontend som setter
        // samme status to ganger lage støy på exchangen.
        if (change.previousStatus() != change.order().getStatus()) {
            events.orderStatusChanged(change.order(), change.previousStatus());
        }
        return change.order();
    }

    private static NotFoundException notFound(UUID orderId) {
        return new NotFoundException("Fant ingen bestilling med id " + orderId + ".");
    }
}

File: 3-APIs/kaffebar-dotnet-solution/Kaffebar/Controllers/OrdersController.cs
using System.ComponentModel.DataAnnotations;
using Kaffebar.Auth;
using Kaffebar.Errors;
using Kaffebar.Events;
using Kaffebar.Models;
using Kaffebar.Storage;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Kaffebar.Controllers;

/// <summary>
/// Bestillinger. Oppgave 1, 2, 5, 7 og 8.
/// </summary>
/// <remarks>
/// MERK — og ikke «rydd» dette bort:
///
/// <c>GET /menu</c> ligger igjen som et Minimal API i Program.cs, mens ordrene bor her
/// i en controller. Det ser inkonsekvent ut, og det er helt med vilje: oppgave 2 ber
/// eksplisitt om nettopp den blandingen, slik at du kan se de to stilene ved siden av
/// hverandre i samme prosjekt og selv vurdere hvor valideringen og
/// OpenAPI-metadataene blir tydeligst. Se FASIT.md, oppgave 2.
///
/// <c>[ApiController]</c> gjør tre ting gratis her:
///   - modellvalidering kjøres automatisk, og returnerer 400 før koden din kjører
///   - parametere bindes fra body/route/query uten <c>[FromBody]</c> og venner
///   - feilsvar blir ProblemDetails
/// </remarks>
[ApiController]
[Route("orders")]
[ProducesResponseType<ProblemDetails>(StatusCodes.Status400BadRequest, "application/problem+json")]
public class OrdersController(IOrderRepository repository, OrderEventPublisher events) : ControllerBase
{
    /// <summary>Hent bestillinger.</summary>
    /// <remarks>
    /// Oppgave 8. Default-verdiene står på parameterne, og havner dermed også i
    /// /openapi/v1.json.
    ///
    /// I et Minimal API ville dette vært <c>[AsParameters] OrderQuery query</c> med en
    /// <c>record OrderQuery(OrderStatus? Status, int Limit = 20, int Offset = 0)</c>.
    /// I en controller er vanlige parametere med default-verdier det som binder
    /// forutsigbart — en record bundet med <c>[FromQuery]</c> får ikke med seg
    /// konstruktørens default-verdier.
    /// </remarks>
    [HttpGet]
    [EndpointSummary("Hent bestillinger")]
    [EndpointDescription("Returnerer bestillinger, nyeste først. Kan filtreres på status og pagineres.")]
    [ProducesResponseType<IEnumerable<Order>>(StatusCodes.Status200OK)]
    public ActionResult<IEnumerable<Order>> ListOrders(
        [FromQuery] OrderStatus? status = null,
        [FromQuery][Range(1, 100)] int limit = 20,
        [FromQuery][Range(0, int.MaxValue)] int offset = 0)
    {
        return Ok(repository.List(status, limit, offset));
    }

    /// <summary>Opprett bestilling.</summary>
    /// <remarks>Oppgave 1. 201 med Location-header — det er det «Created» betyr.</remarks>
    [HttpPost]
    [EndpointSummary("Opprett bestilling")]
    [ProducesResponseType<Order>(StatusCodes.Status201Created)]
    [ProducesResponseType<ProblemDetails>(StatusCodes.Status404NotFound, "application/problem+json")]
    public ActionResult<Order> CreateOrder(CreateOrderRequest request)
    {
        var coffee = repository.FindCoffee(request.CoffeeId);
        if (coffee is null)
        {
            // 404 og ikke 400: kontrakten er overholdt, ressursen finnes bare ikke.
            return ProblemFactory
                .NotFound(HttpContext, $"Fant ingen kaffe med id {request.CoffeeId}. Se GET /menu.")
                .AsResult();
        }

        var order = repository.Create(request, coffee);
        events.OrderCreated(order);

        return Created($"/orders/{order.OrderId}", order);
    }

    /// <summary>Hent én bestilling.</summary>
    /// <remarks>
    /// Oppgave 5 og 6. Route-constrainten <c>:guid</c> er ikke bare dokumentasjon — den
    /// avgjør om ruten matcher i det hele tatt. Se FASIT.md, oppgave 5, om hvorfor
    /// <c>/orders/ikke-en-guid</c> gir 404 her og 400 i Java-sporet.
    /// </remarks>
    [HttpGet("{orderId:guid}")]
    [EndpointSummary("Hent én bestilling")]
    [ProducesResponseType<Order>(StatusCodes.Status200OK)]
    [ProducesResponseType<ProblemDetails>(StatusCodes.Status404NotFound, "application/problem+json")]
    public ActionResult<Order> GetOrder(Guid orderId)
    {
        var order = repository.Find(orderId);
        return order is null ? NotFoundProblem(orderId) : Ok(order);
    }

    /// <summary>Oppdater status på en bestilling.</summary>
    /// <remarks>
    /// Oppgave 7, den anbefalte varianten. PATCH fordi vi oppdaterer én egenskap på en
    /// ressurs som finnes fra før, ikke erstatter hele den.
    /// </remarks>
    [HttpPatch("{orderId:guid}")]
    [Authorize(Policy = AuthPolicies.Barista)]
    [EndpointSummary("Oppdater status på en bestilling")]
    [ProducesResponseType<Order>(StatusCodes.Status200OK)]
    [ProducesResponseType<ProblemDetails>(StatusCodes.Status404NotFound, "application/problem+json")]
    public ActionResult<Order> UpdateOrderStatus(Guid orderId, UpdateOrderStatusRequest request) =>
        ApplyStatus(orderId, request.Status);

    /// <summary>Oppdater status på en bestilling (action-variant).</summary>
    /// <remarks>
    /// Oppgave 7, variant 2. Alias for PATCH. Finnes fordi deltakerne valgte ulikt i
    /// mai, slik at en frontend kan peke på deltakerens eget API uten endringer.
    /// PATCH er anbefalingen — se FASIT.md.
    /// </remarks>
    [HttpPost("{orderId:guid}/status")]
    [Authorize(Policy = AuthPolicies.Barista)]
    [EndpointSummary("Oppdater status på en bestilling (action-variant)")]
    [ProducesResponseType<Order>(StatusCodes.Status200OK)]
    [ProducesResponseType<ProblemDetails>(StatusCodes.Status404NotFound, "application/problem+json")]
    public ActionResult<Order> SetOrderStatus(Guid orderId, UpdateOrderStatusRequest request) =>
        ApplyStatus(orderId, request.Status);

    /// <summary>Én implementasjon, to endepunkter.</summary>
    private ActionResult<Order> ApplyStatus(Guid orderId, OrderStatus status)
    {
        var change = repository.SetStatus(orderId, status);
        if (change is null)
        {
            return NotFoundProblem(orderId);
        }

        // Publiser bare når noe faktisk endret seg. Ellers ville en frontend som setter
        // samme status to ganger lage støy på exchangen.
        if (change.Value.PreviousStatus != change.Value.Order.Status)
        {
            events.OrderStatusChanged(change.Value.Order, change.Value.PreviousStatus);
        }

        return Ok(change.Value.Order);
    }

    private ActionResult<Order> NotFoundProblem(Guid orderId) =>
        ProblemFactory
            .NotFound(HttpContext, $"Fant ingen bestilling med id {orderId}.")
            .AsResult();
}
